libplatsupport/src/plat/star64/serial.c

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2023, UNSW
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <stdlib.h>
+#include <string.h>
+#include <platsupport/serial.h>
+#include <platsupport/plat/serial.h>
+#include "../../chardev.h"
+
+#define UART_THR 0x00 /* UART Transmit Holding Register */
+#define UART_LSR 0x14 /* UART Line Status Register */
+#define UART_LSR_THRE 0x20 /* Transmit Holding Register Empty */
+
+#define REG_PTR(base, off)     ((volatile uint32_t *)((base) + (off)))
+
+int uart_getchar(ps_chardevice_t *d)
+{
+    while ((*REG_PTR(d->vaddr, UART_LSR) & BIT(0)));
+    return *REG_PTR(d->vaddr, UART_THR);
+}
+
+int uart_putchar(ps_chardevice_t *d, int c)
+{
+    if (c == '\n' && (d->flags & SERIAL_AUTO_CR)) {
+        uart_putchar(d, '\r');
+    }
+
+    while ((*REG_PTR(d->vaddr, UART_LSR) & UART_LSR_THRE) == 0);
+
+    /* Add character to the buffer. */
+    *REG_PTR(d->vaddr, UART_THR) = c;
+
+    return c;
+}
+
+static void uart_handle_irq(ps_chardevice_t *dev)
+{
+    /* This UART driver is not interrupt driven, there is nothing to do here. */
```

Comment:
Since the UARTs are used as output only, interrupt support does not matter practically. One day, some generic seL4 console demo (tetris, nethack ...)  would be nice to get read interrupts added also...

Reply:
IIUC based on other examples, to support this the only thing I need to do is to enable the interrupt generation on the device on init, and ack the interrupt to the device in this function?

Reply:
basically yes.

Reply:
I will add that.

But it's not clear to me how a consumer of this API would effectively deal with interrupts. It looks like they are supposed to register and wait for the interrupt themselves, and then tell the driver when the interrupt occurred, without getting any further information about the interrupt from the driver.

On this device there could be various reasons for the interrupt and I don't see a way for a client to determine why it was generated.

Reply:
Added

Reply:
If I remember correctly, uart_handle_irq() is just the abstraction function to ack the interrupt. The lib's user must still attach to the interrupt and call the uart_read() wrapper to drain the FIFO first. An existing app that does this should work out of the box on the Pine64.